selfhost/parser.jakt

```diff
@@ -2166,14 +2160,30 @@ struct Parser {
         let start_span = .current().span()
         .index++
 
-        guard .current() is Identifier(name) else {
-            .error("Expected iterator name", .current().span())
-            return ParsedStatement::Garbage(merge_spans(start_span, .current().span()))
+        mut iterator_name = ""
+        mut destructured_var_decls: [ParsedVarDecl] = []
+        
+        match .current() {
+            Identifier(name) => {
+                iterator_name = name
+                .index++
+            }
+            LParen => {
+                destructured_var_decls = .parse_destructuring_assignment(is_mutable: false)
+                mut tuple_var_name = ""
+                for var in destructured_var_decls.iterator() {
+                    tuple_var_name += var.name
+                    tuple_var_name += "__"
```

Comment:
Uhh...just add a fresh_var() or something?

Reply:
I could add a fresh var, for sure. The way I did was becuase it is easier to understand it in the generated c++ code:

```
Tuple<i64, String> const jakt__idx__word__ = idx__word__;
const i64 idx = ((jakt__idx__word__).get<0>());
 const String word = ((jakt__idx__word__).get<1>());
```

I can change it if it is really necessary

Reply:
if that helped you or you think it helps, sure, let's keep it.

Reply:
I believe this should be any pattern, rather than explicitly checking for `(`.

@alimpfard - for the matching, can we match on a pattern? I'm on a different machine, so it's a little tricky to check. But feels like we should allow the same pattern here we allow with `is` and `match`

Reply:
🤔 I mean if the match is guaranteed to work, why not?
But it does seem rather awkward, what should
```
enum Foo { A, B }
...
for A() in [Foo::A].iterator() {}
```
do here? 
what about
```
enum Foo { A }
...
for A() in [Foo::A].iterator() {}
```

Reply:
@jntrnr I would have to think better about how pattern matching in a `for` statement would work if it were for any pattern instead of just `(`. Maybe investigate how other languages (Rust maybe?) do it.

I would just go ahead and land pattern matching on enums and then build on that on future PRs.

What do you think?

Reply:
fwiw currently the only things allowed after `let` are a name and an open paren for destructuring
I'd suggest putting that in a separate function, and using it here (because ultimately, this for loop binding desugars to a `let`).

Then we wouldn't have to answer these questions in the future, whatever `let` accepts, `for` accepts as well.